### DISCUSSION TITLE
explicit opposite relation error

### DISCUSSION THREAD

**Discussion by mikestecker** [prisma/prisma#5635] [Q&A] [answered]
I'm having an trouble setting up my schema for multiple file relationships.  I keep getting the error `Error validating model "Profile": Automatic related field generation would cause a naming conflict. Please add an explicit opposite relation field.`

```prisma
model Profile {
  ...
  avatar        File @relation("Avatar")
  pastAvatars        File[] @relation("Avatars")
  coverImage        File @relation("CoverImage")
  pastCoverImages        File[] @relation("CoverImages")
  photos        File[] @relation("PhotoGallery")
  ...
}
```

```prisma
model File {
  id                   Int                    @id @default(autoincrement())
  user                 User                   @relation("UserUploadedFile", fields: [userId], references: [id])
  userId               Int
  filename             String
  url                  String                 @unique
  status               ReviewStatus           @default(PENDING_APPROVAL)
  createdAt            DateTime               @default(now())
  updatedAt            DateTime               @updatedAt
}
```

**Comment by ryands17** [comment, marked as answer]
@mikestecker 👋 

In this case, is the Profile to File a 1-many relationship or a many-to-many one?

**Reply by mikestecker** [reply]
Thanks for the reply! It's a 1-many (1 user/profile -> many files): the user/profile can upload many files, but the files are not related to many users.

So the user (through the `Profile` table/model) can upload many `pastAvatars `, `pastCoverImages ` and `photos`. The user would own the files. If the user is deleted, I'd like the files deleted also.

I'll also (eventually) add a user verification service so I'll not only relate the file to the user's profile, but also in a new, future table/model called `Verification`

**Reply by ryands17** [reply]
Whenever you have more than one relation to the same model, you need to specify a relation name. This helps Prisma [disambiguate the relations](https://www.prisma.io/docs/concepts/components/prisma-schema/relations#disambiguating-relations) and assign the foreign keys correctly whenever you add the relation.

So your schema would look something like this:

```prisma
model Profile {
  id              Int    @id @default(autoincrement())
  pastAvatars     File[] @relation("pastAvatars")
  coverImage      File?  @relation("CoverImage")
  pastCoverImages File[] @relation("PastCoverImages")
  photos          File[] @relation("PhotoGallery")
}

model File {
  id                Int      @id @default(autoincrement())
  filename          String
  pastAvatars       Profile? @relation("pastAvatars", fields: [profileId], references: [id])
  coverImage        Profile? @relation("CoverImage", fields: [coverImageId], references: [id])
  pastCoverImages   Profile? @relation("PastCoverImages", fields: [pastCoverImagesId], references: [id])
  photoGallery      Profile? @relation("PhotoGallery", fields: [photoGalleryId], references: [id])
  url               String   @unique
  createdAt         DateTime @default(now())
  updatedAt         DateTime @updatedAt
  profileId         Int?
  coverImageId      Int?
  pastCoverImagesId Int?
  photoGalleryId    Int?
}
```

Let me know if this helps :)

**Reply by mikestecker** [reply]
I see, so the references need to be on both tables? In Prisma 1, I only had to define the relationship on the `Profile` table, not on the `File` table.

**Reply by mikestecker** [reply]
Here's what I ended up with... can you let me know if this looks correct? It seemed to run the migrate process properly without all the extra `id` fields.

```prisma
model User {
  id String @id @default(cuid())

  email     String  @unique
  password String

  // Reverse Relationships
  profile Profile?

  createdAt           DateTime  @default(now())
  modifiedAt          DateTime? @updatedAt
}

model Profile {
  id     String @id @default(cuid())

  // Profile owner
  user   User   @relation(fields: [userId], references: [id])
  userId String @unique

  avatar        File?  @relation(name: "Avatar")
  coverPhotos   File[] @relation(name: "CoverPhotos")
  photos        File[] @relation(name: "PhotoGallery")

  createdAt  DateTime  @default(now())
  modifiedAt DateTime? @updatedAt
}

model File {
  id     String @id @default(cuid())

  // File owner
  user   User   @relation(fields: [userId], references: [id])
  userId String

  filename String
  key      String  @unique
  mimetype String
  encoding String?
  url      String  @unique
  urlThumb String  @unique
  urlSized String  @unique

  // Reverse Relationships to Profile
  profileId                  String?
  profileAvatar              Profile? @relation("Avatar", fields: [profileId], references: [id])
  profileCoverPhotos                Profile? @relation("CoverPhotos", fields: [profileId], references: [id])
  profilePhotoGallery        Profile? @relation("PhotoGallery", fields: [profileId], references: [id])

  createdAt  DateTime @default(now())
  modifiedAt DateTime @updatedAt
}
```

**Reply by ryands17** [reply]
Those extra id's are required as they are the foreign keys to the different types of `File` relations that you have created for the `Profile`. This is similar to modelling a database that if you use the same foreign key, then you would't be able to differentiate which are `photos` and which are `coverPhotos`. I hope that makes sense :)

**Reply by mikestecker** [reply]
Got it. Thank you for all the help explaining how it needs to function. I'm not very strong with databases, and while I still don't fully understand the concepts, this helps out a lot. Thanks again!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
